### PR TITLE
feat: portals market-data

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

This PR provides market-data for portals, i.e:

1. `findAll()` - as mentioned in https://github.com/shapeshift/web/pull/7443#discussion_r1697390648, this is only added for the sake of completeness and as a nice-to-have redundancy
2. `findByAssetId()` i.e the method that's called when navigating to an asset page that doesn't yet have spot market-data
3. `findPriceHistoryByAssetId` i.e historical market-data, fetched when navigating to an asset page depending on the currently selected timeframe

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

- closes https://github.com/shapeshift/web/issues/7305

> What protocols, transaction types or contract interactions might be affected by this PR?

### Asset page spot and historical market-data fetching

- Navigate to an asset for which we don't have initial market-data for, and Coingecko/Coincap doesn't have market-data for e.g `eip155:1/erc20:0xb4e16d0168e52d35cacd2c6185b44281ec28c9dc` / `eip155:43114/erc20:0xc3e6d9f7a3a5e3e223356383c7c055ee3f26a34f`
- Ensure spot market-data looks is fetched when navigating to said asset's page sane against Portal and/or Coingecko/other swappers/explorer
- Ensure historical market-data is fetched and looks sane. This should work for all timeframes, but obviously some assets will still miss it as Portals doesn't have the data
- Ensure `Market Data` section looks sane against Portal and/or Coingecko/other swappers/explorer

### Swapper dispatch of `findByAssetId()`

- Unimplemented yet, see https://github.com/shapeshift/web/pull/7370/files#diff-294e8cf10865c2c47da5f895ee389718c99f1f77468fa92ec2d43b2964f7fcb4R78 for similar logic. TBD as a direct follow-up.

### Runtime testing steps

- ☝🏽 
- Without any prioritization monkey patch, ensure there's no `/v2/tokens/history` XHR spew resulting in rate limited red requests (a normal ~50 requests fetch is expected after wallet unlock)

### `findByAssetId()` fallback testing

**NOTE**: While this method as been implemented for completeness / as a very last resort fallback and works as far as fetching market data goes with the monkey patch below, this *will* spew a bunch of `/v2/tokens/history`, ending up with rate limits when connecting a wallet, as `<AppContext />` will try and dispatch ~2000 concurrent XHRs for assets without market-data. The testing below is just to ensure that the added code works, but happy to remove all its logic and only return an empty object if we believe this is fluff.

- In addition to the above, move `PortalsMarketService` as a very first market-service provider in `src/lib/market-service/market-service-manager.ts`
- Clear your cache fully and do a full page refresh, ideally in `#/connect-wallet` to avoid other requests spewing your console
- Ensure the initial batches of request is succesfully fired and processed, and market-data looks sane across the app for major assets (e.g assets page)

### Operations

- ☝🏽 

## Screenshots (if applicable)

- Spot and historical market-data for Portals assets

https://jam.dev/c/db9c1596-8175-48a5-9d6f-d208974023d9

- `findByAssetId` (NOTE @shapeshift/operations - untestable, this is a monkey-patch for engineering only and should never happen, tl;dr this brings a fallback in case both Coingecko and Coincap were to be down)

image

### CODE REVIEW

File: src/components/AssetHeader/AssetMarketData.tsx
Comment (gomesalexandre) [inline src/components/AssetHeader/AssetMarketData.tsx:131]: Tested with this expensive bad boi and noticed it zero'd out its supply - this makes things better

image
image

File: src/lib/market-service/portals/portals.ts
Comment (gomesalexandre) [inline src/lib/market-service/portals/portals.ts:162]: Added for the sake of completeness and satisfying `MarketService`, though we could just as well return an empty object and call it a day. 

Given the fact that Portals is *after* coingecko and Coincap in the market-service providers, this is realistically never going to be hit.

However, in the unlikely scenario where both Coingecko and Coincap were to be down, we'd currently be left without any market data initial fetch (i.e initial batches) and no subsequent market-data when navigating to asset page either.

This brings a free redundancy path while being a complete implementation of a provider 🎉 

Tested locally after a full cache clear and Portals prioritized over all other providers, and confirmed market-data still looks sane for initial batches fetch.

File: src/lib/market-service/portals/portals.ts
Comment (gomesalexandre) [inline src/lib/market-service/portals/portals.ts:48]: Move me to somewhere shared too

File: src/lib/market-service/market-service-manager.ts
Comment (gomesalexandre) [inline src/lib/market-service/market-service-manager.ts:144]: Rm

File: src/lib/market-service/portals/portals.ts
Comment (gomesalexandre) [inline src/lib/market-service/portals/portals.ts:156]: env

File: src/lib/market-service/portals/portals.ts
Comment (gomesalexandre) [inline src/lib/market-service/portals/portals.ts]: `Math.min(250, count)`

File: src/lib/market-service/portals/portals.ts
Comment (gomesalexandre) [inline src/lib/market-service/portals/portals.ts:179]:
```suggestion
            limit: Math.min(250, count).toString(),
```

File: src/lib/market-service/portals/portals.ts
Comment (gomesalexandre) [inline src/lib/market-service/portals/portals.ts:103]: Note, this guy (`createThrottle` as a singleton) has been redeclared because we can't import from `scripts` - didn't feel like fiddling with tsconfig nor adding a root utils directory as in the scope of this PR

File: src/lib/market-service/market-service-manager.ts
Comment (0xApotheosis) [reply src/lib/market-service/market-service-manager.ts:143]: Why no rm?

File: src/lib/market-service/portals/portals.ts
Comment (0xApotheosis) [inline src/lib/market-service/portals/portals.ts:46]: We already have this in `src/lib/market-service/portals/constants.ts`, yeah?

File: src/lib/market-service/portals/portals.ts
Comment (0xApotheosis) [reply src/lib/market-service/portals/portals.ts:103]: I came here to ask why we weren't re-using `createThrottle` - I found my answer 🙏

File: src/lib/market-service/portals/portals.ts
Comment (0xApotheosis) [reply src/lib/market-service/portals/portals.ts:121]: That's cool, moar redundancy!

**Review by 0xApotheosis — APPROVED** [review]
Approved once those small suggestion actioned 🫡 

✅ No `/v2/tokens/history` `429`s.

Screenshot 2024-07-31 at 4 50 14 PM


✅ Market data on the asset page is very happy

Screenshot 2024-07-31 at 4 45 09 PM

Screenshot 2024-07-31 at 4 47 56 PM

Screenshot 2024-07-31 at 4 48 06 PM

Screenshot 2024-07-31 at 4 48 19 PM

Screenshot 2024-07-31 at 4 48 31 PM

Screenshot 2024-07-31 at 4 48 43 PM

Screenshot 2024-07-31 at 4 48 54 PM


Market data in the swapper is happy

Screenshot 2024-07-31 at 4 45 56 PM

✅ Portals as a fallback provider is happy... until it isn't, as expected

Screenshot 2024-07-31 at 4 53 33 PM

Screenshot 2024-07-31 at 4 55 52 PM

File: src/lib/market-service/market-service-manager.ts
Comment (gomesalexandre) [reply src/lib/market-service/market-service-manager.ts]: 69eb42479c

File: src/lib/market-service/portals/portals.ts
Comment (gomesalexandre) [reply src/lib/market-service/portals/portals.ts]: bda92632

**Comment by gomesalexandre** [comment]
This is ready to go, but moved to draft so we don't merge it into its parent yet until https://github.com/shapeshift/web/pull/7439 is approved and merged)

**Pull request merged**
